theano/contrib/mkl/mkl_type.py

```diff
@@ -0,0 +1,350 @@
+from __future__ import absolute_import, print_function, division
+import os
+import numpy
+from six import StringIO
+
+import theano
+from theano import Type, Variable, config
+from theano.tensor.var import _tensor_py_operators
+
+try:
+    import mkl_ndarray
+    import mkl_ndarray.mkl_ndarray as mkl
+
+    try:
+        mkl_ndarray.__file__
+    except AttributeError:
+        from theano.gof.cmodule import get_lib_extension
+        mkl_ndarray.__file__ = os.path.join(mkl_ndarray.__path__._path[0], 'mkl_ndarray.' + get_lib_extension())
+except ImportError:
+    mkl = None
+
+
+class _operators(_tensor_py_operators):
+    dtype = property(lambda s: s.type.dtype)
+    broadcastable = property(lambda s: s.type.broadcastable)
+    ndim = property(lambda s: s.type.ndim)
+
+
+class MKLNdarrayVariable(_operators, Variable):
+    """
+    See Also
+    --------
+    Variable
+
+    """
+    pass
+
+
+class MKLNdarrayType(Type):
+    """
+    The type that represents an array on CPU with MKL supoorts.
+
+    The `dtype` indicates what scalar data type the elements of
+    variables of this type will be.
+
+    The `broadcastable` indicates whether each dimension is broadcastable
+    or not (to be broadcastable a dimension must always be of length 1)
+
+    The `context_name` is the name of the context. For similar interface
+    with the new back-end for GPU.
+
+    `name` for the type that will be used in printouts.
+
+    """
+    context_name = 'mkl'
+    Variable = MKLNdarrayVariable
+    Constant = None
+    SharedVariable = None
+    ndim = None
+    broadcastable = None
+
+    def __init__(self, dtype, broadcastable, context_name=None, name=None):
+        self.dtype = str(dtype)
+        if self.dtype == 'floatX':
+            self.dtype = config.floatX
+
+        if self.dtype not in ('float32', 'float64'):
+            raise TypeError('%s only supports float32/float64 for now.'
+                            'Tried using dtype %s for variable %s' %
+                            (self.__class__.__name__, dtype, name))
+
+        self.typenum = 11 if dtype is 'float32' else 12
+        self.broadcastable = tuple(bool(b) for b in broadcastable)
+        self.name = name
+        self.dtype_specs()
+        if not (context_name is None):
+            self.context_name = context_name
+
+    def clone(self, dtype=None, broadcastable=None):
```

Comment:
If we need to keep the context_name, I would add it here too for completness. But if you know it isn't useful, don't add it.